Update WIT and binary import/export syntax (#198)
WebAssembly/component-model: design/mvp/WIT.md

[IDL]: https://en.wikipedia.org/wiki/Interface_description_language
[components]: https://github.com/webassembly/component-model

## Package identifiers

All WIT packages are assigned an "ID". IDs look like `foo:bar@1.0` and have

> three components:

Reviewer: I agree that it's useful to have `package` names be optional, in which case all the interfaces being defined by the collection of Wit documents passed to the tool don't have any associated interface ID: it's all kebab names.

> three components:

Author: I agree it would be useful to be able to omit `package` altogether, but that raises enough questions for me that I'm hoping we can get by for now by not implementing it. For example what @lukewagner is proposing means that worlds have drastically different meanings whether or not the enclosing package has an ID or not, and I'm not sure I'm comfortable with that. I would expect some form of opt-in to "yes please interpret everything differently here" more than leaving out `package ...`.

For the JS componentization use case though what you'll be doing is inferring a `world` from a concrete component, and in such a situation there may not have actually ever been a `world` written down in WIT anywhere corresponding to what you're inferring. For example you're inferring a "minimized" version which will be missing functions that aren't otherwise required (e.g. you import `wasi:filesystem/filesystem`'s `stat` function but not `readdir` or something like that). The WIT here, however, entirely synthesized since all you're given was a component. Currently this synthesizes a dummy `package root:root` with a `world root { ... }` in the `Resolve` data structure, since otherwise there's no names to fill in.

Put another way I think this would be good to get to at some point but for now I'm hoping that writing `package foo:foo` at the top or something like that is "good enough".

> three components:

Reviewer 2: @alexcrichton in the above case I would have expected that `wasi:filesystem/filesystem` would still be a direct import of the WASI externname host. It would only be the kebab name types (and implementations) which would be inlined into the component surely?

The fact that we are inferring a `world` from a component is exactly the use case though - componentization is creating components from worlds. Handling that in dependencies is exactly recreating worlds from **these** components. I don't think this use case should be surprising?

I don't fully follow how kebab names break the model in terms of them just being package-local resolutions. I'm fine with some kind of `root:root` pattern to make this stuff work internally, and I understand if you need to move ahead here, my main point is just asking are just:

1. Will these use cases be possible, so we are not cutting them out entirely?
2. Are we not making unnecessary restrictions now that will hinder these use cases unnecessarily now or in future?

> three components:

Reviewer 2: After discussing this further with @alexcrichton, he clarified there may still be ways of supporting kebab name external references in future. But for now that is a secondary use case, since inline types solve the majority of the problem. So effectively:

1. Yes these use cases are possible, but would require further design work in future motivated by real need
2. There are no restrictions being made now that inhibit the externalization of type definitions for kebab names in future

* A namespace, for example `foo` in `foo:bar`. This namespace is intended to
  disambiguate between registries, top-level organizations, etc. For example

>   WASI interfaces use the `wasi:` namespace.

Reviewer: ```suggestion
  WASI interfaces use the `wasi` namespace.
```

>   WASI interfaces use the `wasi:` namespace.

Reviewer: Instead of making namespaces optional, which I think would be a half-way measure, I think these cases should just not use `(interface <id>)` and produce plain kebab-names instead.

> * A package name, for example `bar` in `foo:bar`. This provides a further
>   description of the contained functionality within the namespace. For example

Reviewer: ```suggestion
* A package name, for example `bar` in `foo:bar`. A package name groups together a set
  of interfaces and worlds that would otherwise be named with a common prefix. For example
```

  `wasi:clocks` has the package name `clocks`.

> * An optional version, specified as a major and minor number. Currently full
>   semver isn't allowed and instead only `1.0` is allowed, for example. Both

Reviewer: ```suggestion
* An optional version, specified as a major and minor number. Because WIT packages define
  *interfaces*, not *implementations*, there is not a patch number, as in full semver. Both
```

  version numbers must be unsigned integers. Note that the version field is
  optional, though.

Package identifiers are specified at the top of a WIT file via a `package`
declaration:

```wit
package wasi:clocks
```

WIT packages can be defined in a collection of files and at least one of them
must specify a `package` identifier. Multiple files can specify a `package` and

> they must all agree on what the identifier is.

Author: I'm tempted to start conservatively and require them in the text, but do you have a use case in mind where this would be helpful to have?

> they must all agree on what the identifier is.

Reviewer 2: If we supported the kebab name variant `package foo` that would be fine, but in particular it would be nice to retain the ability to have WIT files that are not part of packages when just using their interfaces and inlining their types, without inlining their names. Ie at the top-level WIT authoring level where you don't even think in terms of a package name yet.

> they must all agree on what the identifier is.

Reviewer 2: After further discussion, the `local:pkg` convention for now will be fine. In future out-of-band configurations eg for kebab names may be possible, depending on how type externalization for them is managed.

[IDL]: https://en.wikipedia.org/wiki/Interface_description_language
[components]: https://github.com/webassembly/component-model

## Package identifiers

All WIT packages are assigned an "ID". IDs look like `foo:bar@1.0` and have
three components:

* A namespace, for example `foo` in `foo:bar`. This namespace is intended to
  disambiguate between registries, top-level organizations, etc. For example
  WASI interfaces use the `wasi` namespace.

* A package name, for example `clocks` in `wasi:clocks`. A package name groups
  together a set of interfaces and worlds that would otherwise be named with a
  common prefix.

* An optional version, specified as a major and minor number. Because WIT
  packages define *interfaces*, not *implementations*, there is not a patch

>   number, as in full semver. Both version numbers must be unsigned integers.

Reviewer: I did realize a potential use case for including an optional patch version: when you're iterating on a pre-1.0 interface, and minor version bumps can be breaking, then you might want to use patch version changes to signal non-breaking interface changes.  Concretely, I could imagine that for Preview 2, we'd want to publish `0.` versions of WASI interfaces (say, `0.2`), and then if we make non-breaking changes post-Preview 2, we could publish them as `0.2.x` versions (saving `0.3` for Preview 3).

Reviewer 3: We could also allow a pre-release version number and potentially a build number. I think this would more closely match the actual semantics (i.e., it's not a patch; it's a pre-release)

Reviewer 4: In the past we had discussed using patch versions for non-functional changes. Perhaps the only, but still important, change of this kind would be documentation changes. I think those warrant the inclusion of patch version numbers

Author: Ok sounds reasonable to me, I went ahead and added full semver in including pre-release/build metadata.

  Note that the version field is optional, though.

Package identifiers are specified at the top of a WIT file via a `package`
declaration:

```wit
package wasi:clocks
```

or

```wit
package wasi:clocks@1.2
```

WIT packages can be defined in a collection of files and at least one of them
must specify a `package` identifier. Multiple files can specify a `package` and

> they must all agree on what the identifier is.

Author: @guybedford do you have thoughts on this?

> they must all agree on what the identifier is.

Reviewer 2: I'd be open to changing the constraint to a single file, but see no problem with the redundancy either.

> they must all agree on what the identifier is.

Reviewer 2: Also the concept of `main` doesn't quite apply to WIT currently since all WIT files are equal. Unless we introduce preferential lookup rules for default worlds and interfaces.

> they must all agree on what the identifier is.

Reviewer 3: There are other tools that have this convention. For example, terraform has the convention of main files that (I believe) have no significance other than to say to the reader "look here first".

An example of an interface is:

```wit

> package local:demo

Author: @guybedford you might have thoughts on this? (I don't mind either way myself)

> package local:demo

Reviewer 2: My concern was exactly that asking users to choose an arbitrary namespace at an early stage risks exactly the naming problem namespaces are supposed to avoid in the first place.

> package local:demo

Reviewer 3: So does `local` actually have a meaning here or is this a convention you're trying to establish or is this purely descriptive? Either way, it would be nice to explicitly state that in this document.

and [function][functions] definitions. For example:

```wit

> package wasi:filesystem
> 
> interface filesystem {
>   use types.{errno}

Reviewer: ```suggestion
package wasi:filesystem

interface types {
  use wasi:clocks.wall-clock.{datetime}
```
The goal here being to have this example slightly more-closely align with the real `wasi:filesystem` which wants to define the interface ID `wasi:filesystem/types`.  If this change makes sense, you'd probably stick a `datetime` inside the `stat` record and remove `errno` from the `result` of `stat-file`.

function or an interface.

```wit
package wit:demo

world command {
  import wasi:filesystem/filesystem
  import wasi:random/random
  import wasi:clocks/monotonic-clock
  // ...

  export main: func(args: list<string>)
}
```

More information about the `wasi:random/random` syntax is available below in the
description of [`use`][use].

An imported or exported interface corresponds to an imported or exported
instance in the component model. Functions are equivalent to bare component
functions.
Additionally interfaces can be defined inline with an explicit kebab-name that
avoids the need to have an out-of-line definition.

```wit
package wit:demo

interface out-of-line {
  the-function: func()
}

world your-world {
  import out-of-line
  // ... is roughly equivalent to ...
  import out-of-line: interface {
    the-function: func()
  }

> }

Reviewer 2: I get this framing, and it's a nice approach, but I wasn't aware we were going to entirely separate kebab names as no longer being able to reference other dependencies at all.

Again for the componentization use case, if I want to target a dependency that is another component wasm file, it would be nice to be able to have tooling that could do eg:

```
world parent-world {
  import dep
}
```

and then be able to define `dep` to a local component when componentizing:

```
jco componentize x.js --world parent-world --dep dep=./child.wasm
```

It would be nice to leave the door open to interesting use cases like the above.

Reviewer 2: > If we go via the route described above where this form automatically uses kebab-names in some situations, I think that would satisfy the use case here?

Can you clarify what you mean by this?

```

The kebab name of the `import` or `export` is the name of the corresponding item

> in the final component. This can be different from the [`interface`][interfaces]
> name but must be a [valid identifier][identifiers]. Note that kebab names cannot

Reviewer: ```suggestion
in the final component. Note that kebab names cannot
```

overlap and must be unique, even between imports and exports.

Note that in the component model imports to a component either use an ID or a
kebab-name, and in WIT this is reflected in the syntax:

```wit
package wit:demo

interface my-interface {
  // ..
}

world command {
  // generates an import of the ID `wit:demo/my-interface`
  import my-interface

>   // generates an import of the ID `wasi:random/random`
>   import wasi:random/random

Reviewer: ```suggestion
  // generates an import of the ID `wasi:filesystem/types`
  import wasi:filesystem/types
```

Reviewer: (To avoid the confusion of "why are there two `random`s?".)

function or an interface.

```wit
package local:demo

world command {
  import wasi:filesystem/filesystem
  import wasi:random/random
  import wasi:clocks/monotonic-clock
  // ...

  export main: func(args: list<string>)
}
```

More information about the `wasi:random/random` syntax is available below in the
description of [`use`][use].

An imported or exported interface corresponds to an imported or exported
instance in the component model. Functions are equivalent to bare component
functions.
Additionally interfaces can be defined inline with an explicit kebab-name that
avoids the need to have an out-of-line definition.

```wit
package local:demo

interface out-of-line {
  the-function: func()
}

world your-world {
  import out-of-line
  // ... is roughly equivalent to ...
  import out-of-line: interface {
    the-function: func()
  }
}
```

The kebab name of the `import` or `export` is the name of the corresponding item

> in the final component. Note that kebab names cannot overlap and must be unique,

Author: Good point! I've clarified that this use case is allowed.

```wasm
(component
  (import (interface "wit:demo/shared") (instance $shared
    (type $metadata (record (; ... ;)))
    (export "metadata" (type (eq $metadata)))
  ))
  (alias export $shared "metadata" (type $metadata'))
  (import "host" (instance $host
    (export "metadata" (type $metadata (eq $metadata')))
    (export "get" (func (result $metadata)))
  ))
)
```

Here it can be seen that despite the `world` only listing `host` as an import
the component additionally imports a `wit:demo/shared` interface. This is due

> to the fact that the `use { ... } from shared` implicitly requires that

Reviewer: ```suggestion
to the fact that the `use shared.{ ... }` implicitly requires that
```

```wasm
(component
  (type $demo (component
    (type $console (instance
      (export "log" (func (param "arg" string)))
    ))
    (export (interface "wit:demo/console") (instance (type $console)))
  ))

>   (export (interface "wit:demo/wit") (type $demo))

Reviewer: I'm a bit confused where the final `/wit` comes at the end of `wit:demo/wit`.  If we want to preserve file grouping, I'd expect the ID to be `wit:demo/host` or even just the kebab-name `host` (highlighting the fact that `host` is not an *interface*).  If we don't want to preserve file grouping, then maybe export the whole package with the kebab-name of the package (again, because it's not an *interface*).

Author: Good question! This is here though not for interfaces but instead for the package itself. I copy-pasted this output from `wasm-tools component wit` which is why it's here, but my thinking is that the package name should be recorded somewhere, probably around here, but IDs require a `/foo` and when naming the package there's nothing to put there, so I arbitrarily put in `/wit`.

This doesn't actually reflect file structure though, it's only there to satisfy the validation requirements of an interface name being present. All contents of the package are in the component type itself.

Basically this ideally would be simply `"wit:demo"`, but that doesn't validate.

Reviewer: Ah, hah, gotcha.  One observation is that, since the namespace, package name and version are already present inside each exported interface and world ID, there's no additional info provided by this.  In fact, I wonder if we even need this enclosing component-type at all: could the outermost component simply contain an *unnamed* `type`-definition for each interface-or-world which is a component-type that imports `use`d dependencies and exports the interface-or-world, establishing the ID via `(export (interface ...) ...)`?

Author: Validation-wise, yes, that should work. Integration-wise though that would be more difficult I think though because index spaces are generally considered to be implementation details of wasms where the exports/imports define how the embedder interacts with it. 

One other issue though is that if a package is empty with no other interfaces or worlds I don't think the ID would be recorded anywhere?

Reviewer: I suppose another slight variation is for each interface/world to be a separate type definition that is exported with the ID (which would then be always be duplicated).  I guess mainly I was realizing that it doesn't seem like the component-types representing each individual interface/world need to be wrapped up into a single component type.

It's true that if the package was empty, there wouldn't be any record of the package name, but it seems like that's fine, since there's nothing to refer to.

Author: I did actually initially go for the containerless format like you're mentioning. The names actually already duplicate the package name as each interface is exported with its full ID rather than just its name. The problem is that without a container you're generating an actual component with actual imports since interfaces may need to import interfaces they depend on. That would change things radically since this is supposed to be "just" type-level things.

Reviewer: Good point, I was misreading the current output and you're right.  Thinking about it some more, given the following Wit:
```wit
package wasi:http
interface types {
  resource request { ... }
  resource response { ... }
}
interface handler {
  use types.{request, response}
  handle: func(request) -> response
}
world proxy {
  import handler
  export handler
}
```
I think perhaps we'd want to get the following Wat encoding:
```wat
(component
  (component (export "types")
    (type (export (interface "wasi:http/types")) (instance
      (export "request" (type (sub resource)))
      (export "response" (type (sub resource)))
    ))
  )
  (component (export "handler")
    (import (interface "wasi:http/types") (instance $types
      (export "request" (type (sub resource)))
      (export "response" (type (sub resource)))
    ))
    (alias export $types "request" (type $request))
    (alias export $types "response" (type $response))
    (type (export (interface "wasi:http/handler")) (instance
      (export "handle" (func (param "request" (own $request)) (result (own $response))))
    ))
  )
  (component (export "proxy")
    (type (export (interface "wasi:http/proxy")) (component
      (import (interface "wasi:http/types") (instance $types
        (export "request" (type (sub resource)))
        (export "response" (type (sub resource)))
      ))
      (alias export $types "request" (type $request))
      (alias export $types "response" (type $response))
      (import (interface "wasi:http/handler") (instance
        (export "handle" (func (param "request" (own $request)) (result (own $response))))
      ))
      (export (interface "wasi:http/handler") (instance
        (export "handle" (func (param "request" (own $request)) (result (own $response))))
      ))
    ))
  ))
)
```
Here, each interface and world is individually wrapped in a component that is able to express `use` dependencies via imports.  For worlds, there's currently never a need to have imports in the wrapper, but I'm not 100% sure this would never make sense in the future and it seems nice to have interfaces and worlds be symmetric, distinguished by whether the wrapper component exports an `instance` type (for an interface) or a `component` type (for a world).

A nice thing about this encoding is that it makes it clear that each interface/world definition is a standalone unit and thus packages can be arbitrarily split and merged.  It also makes clear that each interface's `use`s can be resolved independently and/or in multiple ways, depending on context, which is less clear with the current encoding where they alias each other directly.

Another nice thing is that, given an interface ID like `wasi:http/handler`, you can think of the `wasi:http` as mapping to the above component package and then the trailing `handler` projects out the `handler` subcomponent which is also how nested component *implementation* packages would work.  This also naturally generalizes to deeper nesting (i.e., `foo:bar/baz/quux`) if we ever wanted that.  Also, if we ever wanted to allow `foo:bar` to refer to a specific interface or world, we could achieve that by having the outermost component contain `(type (export (interface "foo:bar")) ...)`, which we could do while still allowing there to be interfaces/worlds named `foo:bar/baz`.  (E.g., `wasi:http` could map to what's currently called `wasi:http/proxy`, if we wanted.)

WDYT?

Author: I suspect something along these lines could work, but there's two issues given the current translation as I understand it:

1. An empty wit package would be translated as `(component)` which doesn't include the package's name and it's also ambiguous with an acutal empty component itself. While it's sort of an edge case this is all so low-level I'd ideally prefer to avoid any accidental ambiguities with things like this.
2. Do exports have to be of actual concrete components? An operation supported today is "extract the WIT from this concrete component" and that transparently works with actual components and wasm-encoded WIT packages. The detection is currently "does this component export a component type" since no real component will realistically do that for a long time. This detection, however, gets more complicated as the structure of the component looks "real" until you get to the exported instance/component types. Additionally I personally thought it was a nice feature to encode the entire WIT package in the type section, where this is sort of half actual runtime things half type things.

Reviewer: Yeah, both great questions.  I'm not sure these are the absolute best answers, but I'll put them out there to see what you think:
1. In a full workflow where this package gets published, the registry to which it was published would establish the name.  We've also talked about a registry-defined package metadata custom section that would contain information like this too.  I suppose why it seems fine to have a different answer here is that packages are not "semantic" things with IDs in the same way that interfaces and worlds are.
2. I could see s/component def/component type/ working too, almost equivalently, so I've flip-flopped in my head on this one.  For rendering `.wit` from `.wasm`, I think the "signal" would be any `(export (interface ...) (type ...))` and so the Wit-rendering tool would recurse over the component tree and splat out each of these it found into the generated `.wit` document.  As a nice side effect, you could then bundle `.wit` with *any* component `.wasm`.  The other small neat thing about using component defs as the "wrapper" is that last little thing I mentioned about being able to have the outermost component serve as the wrapper (possibly, at some point in the future).

Author: Given how low-level this tooling is though I'd ideally prefer to not require or assume a registry in place. Basically otherwise everything will need to carry the name around manually separate from the wasm itself, whereas otherwise it's stored inline in the wasm with the current proposal.

I'm still not sure I understand the "neat" part here though myself. Everything feels far enough in the future that nothing is really solid enough that we would be able to predict that an alternate design would indeed fit neatly into a design one way or another. I'm hesitant to change designs now radically for a proposed future design which isn't fully fleshed out.

Reviewer: Regarding the registry aspect: sorry, I didn't mean to imply that the package name lives in the registry, but rather a custom section in the component binary that could also hold other package-metadata types of fields.

That being said, in the short term, maybe it's better to just make the minimal change to the existing encoding to unblock other stuff.  Because we'll eventually be publishing these Wit packages to a registry, the encoding scheme will need to be effectively standardized to allow tools to interoperate and then it'll be important to think long-term about some of these packaging questions, but we're not there yet.
